timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampLockCollection.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.lock;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import java.util.Optional;
+
+final class NamedMinTimestampLockCollection {
+    private static final String IMMUTABLE_TIMESTAMP_NAME = "ImmutableTimestamp";
+
+    private final LoadingCache<String, NamedMinTimestampTracker> namedMinTimestampTrackers =
+            Caffeine.newBuilder().build(NamedMinTimestampTracker::new);
+
+    AsyncLock getImmutableTimestampLock(long timestamp) {
+        return getNamedMinTimestampLockInternal(IMMUTABLE_TIMESTAMP_NAME, timestamp);
+    }
+
+    Optional<Long> getImmutableTimestamp() {
+        return getNamedMinTimestampInternal(IMMUTABLE_TIMESTAMP_NAME);
+    }
+
+    private AsyncLock getNamedMinTimestampLockInternal(String name, long timestamp) {
+        return NamedMinTimestampLock.create(getNamedMinTimestampTracker(name), timestamp);
```

Comment:
Let's throw if name equals "ImmutableTimestamp?

Reply:
We do not need to explicitly throw. I am using the API object that does the validation as recommended and I pass it as deep as possible.